src/persist/src/operators/mod.rs

```diff
@@ -24,3 +39,125 @@ pub(crate) fn split_ok_err<K, V>(
+    /// follows the following pattern:
+    ///
+    /// ```ignore
+    /// op.build_async(scope, move |capabilities, frontier, scheduler| async move {
+    ///     while scheduler.yield_now().await {
+    ///         // async operator logic here
+    ///     }
+    /// });
+    /// ```
+    ///
+    /// Since timely's input handles and frontier notifications are not integrated with the async
+    /// ecosystem the only way to yield control back to timely is by awaiting on
+    /// `scheduler.yield_now()`. The operator will ensure that this call resolves when there is
+    /// more work to do.
+    fn build_async<B, Fut>(self, scope: G, constructor: B)
+    where
+        B: FnOnce(
+            Vec<Capability<G::Timestamp>>,
+            Rc<RefCell<Vec<Antichain<G::Timestamp>>>>,
+            Scheduler,
+        ) -> Fut,
+        Fut: Future + 'static;
+}
+
+impl<G: Scope> OperatorBuilderExt<G> for OperatorBuilder<G> {
+    fn build_async<B, Fut>(self, scope: G, constructor: B)
+    where
+        B: FnOnce(
+            Vec<Capability<G::Timestamp>>,
+            Rc<RefCell<Vec<Antichain<G::Timestamp>>>>,
+            Scheduler,
+        ) -> Fut,
+        Fut: Future + 'static,
+    {
+        let activator = scope.sync_activator_for(&self.operator_info().address[..]);
+        let waker = futures_util::task::waker(Arc::new(activator));
+        let shared_frontiers = Rc::new(RefCell::new(vec![
+            Antichain::from_elem(Timestamp::minimum());
+            self.shape().inputs()
+        ]));
+
+        self.build(move |capabilities| {
+            let scheduled = Rc::new(Cell::new(false));
+            let mut logic_fut = Box::pin(constructor(
+                capabilities,
+                Rc::clone(&shared_frontiers),
+                Scheduler(Rc::clone(&scheduled)),
+            ));
+            move |frontiers| {
+                // Attempt to update the shared frontier before polling the future.  This operation
+                // can fail if the future also borrowed the frontier and kept the borrow active
+                // across an await point. It is fine for this operation to fail because we will
+                // poll the future right afterwards and will come back to this point once the
+                // operator gets rescheduled. At that future moment we will be able to update the
+                // frontier and poll the future with fresh data in the handle.
+                if let Ok(mut shared_frontiers) = shared_frontiers.try_borrow_mut() {
+                    for (shared, new) in shared_frontiers.iter_mut().zip(frontiers) {
+                        if !PartialOrder::less_equal(&new.frontier(), &shared.borrow()) {
+                            *shared = new.frontier().to_owned();
+                        }
+                    }
+                }
+
+                let was_marked_scheduled = scheduled.replace(true);
+                let _ = Pin::new(&mut logic_fut).poll(&mut Context::from_waker(&waker));
```

Comment:
this only returns a Ready if the logic violates the contract ("The logic constructror is expected to return a future that will never return") right? it might be worth logging or debug_assert'ing or something to make it more obvious what is going on if someone accidentally makes a change that violates that contract